clients/src/main/java/org/apache/kafka/common/serialization/Deserializer.java

```diff
@@ -32,13 +32,13 @@
     public void configure(Map<String, ?> configs, boolean isKey);
     
     /**
-     *
+     * Deserialize a record value from a bytearray into a value or object.
      * @param topic topic associated with the data
-     * @param data serialized bytes
-     * @return deserialized typed data
+     * @param data serialized bytes; may be null. If null, implementations should return null.
```

Comment:
Oh yeah… technically and conceptually, that makes sense, I think I agree.

Some context: I was (erroneously) approaching this specifically from the point of view of Kafka Streams aggregations, wherein it’s highly feasible for no prior data to be found, at which point the framework will currently pass a null to a Deserializer and expect the Deserializer to return a value — as opposed to throwing an exception. (My Deserializer wasn’t expecting nulls and therefore threw an NPE.)

I [discussed this](http://mail-archives.apache.org/mod_mbox/kafka-users/201606.mbox/%3cCAHwHRrU29jw4jgVhSijWbVLZB3BC6Qz6pbH9TqCFBcoRJk4aZQ@mail.gmail.com%3e) on the kafka-users list with @guozhangwang and he wrote “I think in general the library should guard this case instead of letting desers worry about it.” I agreed so I opened [KAFKA-3836](https://issues.apache.org/jira/browse/KAFKA-3836) as he suggested.

I guess what I really needed to know in this case was that `deserialize` needs to be able to handle null input. It does seem that I went too far with “implementations should return null.” I’ll remove it from this class. That guidance _might_ belong somewhere in Streams. Maybe in particular on aggregation methods such as `reduceByKey`. @guozhangwang what do you think?